java/ql/src/experimental/CWE-918/RequestForgeryCustomizations.qll

```diff
@@ -0,0 +1,182 @@
+/** A module to reason about request forgery vulnerabilities. */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.frameworks.ApacheHttp
+import semmle.code.java.frameworks.spring.Spring
+import semmle.code.java.frameworks.JaxWS
+import semmle.code.java.frameworks.javase.Http
+import semmle.code.java.dataflow.DataFlow
+
+/** A module to reason about request forgery vulnerabilities. */
+module RequestForgery {
+  /** A data flow sink for request forgery vulnerabilities. */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * An argument to an url `openConnection` or `openStream` call
+   *  taken as a sink for request forgery vulnerabilities.
+   */
+  private class UrlOpen extends Sink {
+    UrlOpen() {
+      exists(MethodAccess ma |
+        ma.getMethod() instanceof UrlOpenConnectionMethod or
+        ma.getMethod() instanceof UrlOpenStreamMethod
+      |
+        this.asExpr() = ma.getQualifier()
+      )
+    }
+  }
+
+  /**
+   * An argument to an Apache `setURI` call taken as a
+   *   sink for request forgery vulnerabilities.
+   */
+  private class ApacheSetUri extends Sink {
+    ApacheSetUri() {
+      exists(MethodAccess ma, TypeApacheHttpRequestBase t |
+        ma.getReceiverType().extendsOrImplements(t) and
```

Comment:
Why the `.extendsOrImplements`? `TypeApacheHttpRequestBase` already includes subclasses.